channeldb/circuit.go

```diff
@@ -0,0 +1,172 @@
+package channeldb
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"encoding/hex"
+	"github.com/boltdb/bolt"
+	"github.com/roasbeef/btcd/wire"
+)
+
+var (
+	// circuitBucket is the name of the bucket within the database that
+	// stores all data related to circuits.
+	circuitBucket = []byte("circuits")
+)
+
+// CircuitKey uniquely identifies an active Sphinx (onion routing) circuit
+// between two open channels. Currently, the rHash of the HTLC which created
+// the circuit is used to uniquely identify each circuit.
+type CircuitKey [sha256.Size]byte
+
+func (k *CircuitKey) String() string {
+	return hex.EncodeToString(k[:])
+}
+
+// PaymentCircuit is used by HTLC switch service in order to determine
+// backward path for settle/cancel HTLC messages. A payment circuit is created
+// once a htlc manager forwards an HTLC add request. Channel points contained
+// within this message is used to identify the source/destination HTLC managers.
+//
+// NOTE: In current implementation of HTLC switch, the payment circuit might be
+// uniquely identified by payment hash but in future we implement the payment
+// fragmentation which makes possible for number of payments to have
+// identical payments hashes, but different source and destinations.
+//
+// For example if Alice(A) want to send 2BTC to Bob(B), then payment will be
+// split on two parts and node N3 will have circuit with the same payment hash,
+// and destination, but different source (N1,N2).
+//
+//	    	  1BTC    N1   1BTC
+//    	      + --------- o --------- +
+//      2BTC  |	                      |  2BTC
+// A o ------ o N0	           N3 o ------ o B
+//	      |		              |
+// 	      + --------- o --------- +
+//	         1BTC     N2   1BTC
+//
+type PaymentCircuit struct {
+	// PaymentHash used as uniq identifier of payment (not payment circuit)
+	PaymentHash CircuitKey
+
+	// Src is the channel id from which add HTLC request is came from and
+	// to which settle/cancel HTLC request will be returned back. Once the
+	// switch forwards the settle message to the source the circuit is
+	// considered to be completed.
+	Src *wire.OutPoint
```

Comment:
Rather than storing the `*wire.OutPoint` here, we can instead just store the value directly instead of pointing to it. This can at times improve cache locality and will also reduce GC pressure as the entire object can be GC'd more quickly than otherwise.